service/node.go

```diff
@@ -504,9 +519,126 @@ func (s *service) NodeGetInfo(
+
+	csiVolID := req.GetVolumeId()
+	volPath := req.GetVolumePath()
+	mounted := false
+	healthy := true
+	message := ""
+
+	//validate params first, make sure neither field is empty
+	if len(volPath) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "no volume Path provided")
+	}
+
+	if len(csiVolID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "no volume ID  provided")
+	}
+
+	//check if volume exists
+
+	volID := getVolumeIDFromCsiVolumeID(csiVolID)
+	systemID := s.getSystemIDFromCsiVolumeID(csiVolID)
+
+	if systemID == "" {
+		// use default system
+		systemID = s.opts.defaultSystemID
+	}
+
+	_, err := s.getSDCMappedVol(volID, systemID, 30)
+	if err != nil {
+		//volume not known to SDC, next check if it exists at all
+		_, _, err := s.listVolumes(systemID, 0, 0, false, false, volID, "")
+		if err != nil && strings.Contains(err.Error(), sioGatewayVolumeNotFound) {
+			message = fmt.Sprintf("Could not get volume with ID '%s' from array %s due to error: %s", volID, systemID, err)
+
+		} else if err != nil {
+			//err wasn't not found, return error
+			return nil, err
+		}
+		//volume was found, but was not known to SDC. This is abnormal.
+		healthy = false
+		if message == "" {
+			message = fmt.Sprintf("volume: %s was not mapped to host: %v", volID, err)
+		}
+
+	}
+
+	//check if volume path is accessible
+	if healthy {
+		_, err = os.ReadDir(volPath)
+		if err != nil && healthy {
+			healthy = false
+			message = fmt.Sprintf("volume path: %s is not accessible: %v", volPath, err)
+		}
+	}
+
+	if healthy {
+
+		//check if path is mounted on node
+		mounts, err := getPathMounts(volPath)
+		if len(mounts) > 0 {
+			for _, m := range mounts {
+				if m.Path == volPath {
+					Log.Infof("volPath: %s is mounted", volPath)
+					mounted = true
+				}
+			}
+		}
+		if len(mounts) == 0 || mounted == false || err != nil {
+			healthy = false
+			message = fmt.Sprintf("volPath: %s is not mounted: %v", volPath, err)
+		}
+
+	}
+
+	if healthy {
+
+		availableBytes, totalBytes, usedBytes, totalInodes, freeInodes, usedInodes, err := k8sutilfs.Info(volPath)
+		if err != nil {
+			return &csi.NodeGetVolumeStatsResponse{
+				VolumeCondition: &csi.VolumeCondition{
+					Abnormal: false,
+					Message:  fmt.Sprintf("failed to get metrics for volume with error: %v", err),
+				},
+			}, nil
+
+		}
+		return &csi.NodeGetVolumeStatsResponse{
+			Usage: []*csi.VolumeUsage{
+				{
+					Available: availableBytes,
+					Total:     totalBytes,
+					Used:      usedBytes,
+					Unit:      csi.VolumeUsage_BYTES,
+				},
+				{
+					Available: freeInodes,
+					Total:     totalInodes,
+					Used:      usedInodes,
+					Unit:      csi.VolumeUsage_INODES,
+				},
+			},
+			VolumeCondition: &csi.VolumeCondition{
+				Abnormal: !healthy,
+				Message:  message,
+			},
+		}, nil
+
+	}
+
+	return &csi.NodeGetVolumeStatsResponse{
+		VolumeCondition: &csi.VolumeCondition{
```

Comment:
may have to send empty usage object , it cannot be nil

Reply:
added

Reply:
good work , you should see events in pod